File: scm-java/src/main/scheme/primitives/PrimitivePairSource.java
package scheme;

public class PrimitivePairSource extends Primitive {
    @Override public String name() { return "%pair-source"; }

    @Override public String info() {
        return "Syntax: (%pair-source pair)\n" +
               "Library: (scm core)\n" +
               "Description: Returns the source position of a pair as (filename . line), or #f if no position is available.";
    }

    @Override public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 1, 1);
        if (Value.isPair(arguments[0])) {
            SourcePos p = Value.asPair(arguments[0]).pos;
            if (p != null)
                return new Pair(p.filename.toCharArray(), p.line);
        }
        return Value.F;
    }
}

File: scm-csharp/scheme/primitives/PrimitivePairSource.cs
namespace scheme;

public class PrimitivePairSource : Primitive
{
    public override string Name() => "%pair-source";

    public override string Info() =>
        "Syntax: (%pair-source pair)\n" +
        "Library: (scm core)\n" +
        "Description: Returns the source position of a pair as (filename . line), or #f if no position is available.\n" +
        "Example:\n" +
        "  (%pair-source '(a b)) => #f";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 1, 1);
        if (Value.IsPair(arguments[0]))
        {
            var p = Value.AsPair(arguments[0]).pos;
            if (p != null)
                return new Pair(p.filename.ToCharArray(), p.line);
        }
        return Value.F;
    }
}
